### DISCUSSION TITLE
Database Array Explode/Implode

### DISCUSSION THREAD

**Discussion by LaocheXe** [e107inc/e107#4673] [Q&A Plugins]
I've been wondering about this for years now, and have asked around but still I don't understand how this works.

I have a service records plugin, which stores data of users, some of the details stored are in an array (kind of like userclasses in the user database), It saves unique id's from other tables. which would end up looking like (1,4,15,22, etc...).

I want to show the details in the user profile, however I can't wrap my head around it. explode/implode - if I explode the array, how would I link the numbers(id's) to the other database to return their name, or other information?

Since it's similar to userclasses, I've been digging around the code and end up null on my end.

---- Alright, time for a better example. I think....

So I have a list of "Advanced Clone Training" <--- name doesn't really matter for this - With this is a list of "Trainings" that a unit would have created for their members to do, etc....

![Screenshot 2022-01-24 080243](https://user-images.githubusercontent.com/5818022/150797809-d7d744fb-ae60-482b-8e60-336484b27b46.png)

In the database is looks like this:
![Screenshot 2022-01-24 080510](https://user-images.githubusercontent.com/5818022/150798088-f4eca605-a191-49ec-89a2-882c531c6e01.png)

Now for the unit, there is a service records area, which allows administrators to edit a member's service record to keep it up to date, etc...

![Screenshot 2022-01-24 080357](https://user-images.githubusercontent.com/5818022/150798263-1396f237-9bc3-4aab-ada6-6905a93ce0d6.png)

That is stored as:

![Screenshot 2022-01-24 080612](https://user-images.githubusercontent.com/5818022/150798319-bfed9321-826e-4a0a-83c0-845274aa57d9.png)

What I'd want to do is display that service record information on the user profile page. However, I'm use to do it like Jimmi08 said, in one query (but I know it can't be done, this is why I asked).

**Comment by Deltik** [comment]
e107 is currently not normalizing data correctly, so I recommend against storing data like how e107 saves `e107_user.user_class`.  According to the [first normal form](https://en.wikipedia.org/wiki/First_normal_form) (1NF), database table cells may not store an array data type.

Suppose you have an Author and many Books. To model that a Book _belongs to_ an Author, you add a column `books.author_id` with a foreign key relationship to `authors.id` like this:

![image](https://user-images.githubusercontent.com/1364268/150670851-05f7786c-a739-4142-9a83-6feb307edbc2.png)
<sup><sub>Source: https://guides.rubyonrails.org/association_basics.html#the-belongs-to-association</sub></sup>

Author then _has many_ books:

![image](https://user-images.githubusercontent.com/1364268/150670891-6f117b2a-1587-4db4-accd-d3bc975feeee.png)
<sup><sub>Source: https://guides.rubyonrails.org/association_basics.html#the-has-many-association</sub></sup>

But what if you wanted to connect two tables together so that they each have many of each other?  A Physician _has many_ Patients _through_ Appointments, and a Patient also _has many_ Physicians _through_ Appointments.  Your database schema should look like this, then:

![image](https://user-images.githubusercontent.com/1364268/150671038-1dd08ffd-5609-41ef-9efe-b1653c137b6f.png)
<sup><sub>Source: https://guides.rubyonrails.org/association_basics.html#the-has-many-through-association</sub></sup>

To fix e107 userclasses, a third table, `e107_user_userclasses`, should be created, and there should be columns like:
* `e107_user_userclasses.id` – Primary key
* `e107_user_userclasses.user_id` – Foreign key to `e107_user.user_id`
* `e107_user_userclasses.userclass_id` – Foreign key to `e107_userclass_classes.userclass_id`

This model would mean that an `e107_user` has many `e107_userclass_classes` through `e107_user_userclasses` and that an `e107_user_userclasses` has many `e107_user`s through `e107_user_userclasses`, and this model would comply with 1NF.

**Reply by Jimmi08** [reply]
@LaocheXe you are using userclasses as bad example - and it is the reason for the deltic answer.

In your case you have a field with multiple values and options for this field are saved in a separate table.  

For displaying just use explode + foreach values get displayed the needed value and implode them back to one string. 

For searching use FIND IN SET

You are trying to get everything in one query, which is too complicated.

**Reply by LaocheXe** [reply]
So I would use a simple query for the Service records db, then explode the field with multiple id's, then foreach with a FIND IN SET query from the other db table?

**Reply by Jimmi08** [reply]
No, FIND IN SET  is used when you need all users with that value x and value is part of string.   
`https://www.w3schools.com/sql/func_mysql_find_in_set.asp`
 

Real example:

$charlist - array of ids and values;
$characters -  user value (string of 1,2,3...)
$charlinks - result;

```
$characters_array = explode(",", $characters);
$charlinks = array( );

foreach($characters_array as $c) {
        if(empty($charlist[$c]['name'])) continue;
        if(***)   $charlinks[] = "<a href='"***'>".$charlist[$c]['name']."</a>";
        else $charlinks[] = $charlist[$c]['name'];
 }
return implode(", ", $charlinks);
```
You don't need array and implode(), normal string is enough too.

**Reply by LaocheXe** [reply]
Question, what is the ** for the if(***)? Trying to follow whats there, and I might be going at it wrong.

`
$actlist = $sql->retrieve('act_progress_sys', '*'); // Returns array
		$actuprogress = $sr['actp_id']; // Returns actp id's that the user has in service records
		$actuprogress_array = explode(',', $actuprogress);
		$actplink = array( );

foreach($actuprogress_array as $act)
		{
			if(empty($actlist[$act]['actp_name'])) continue;
			if(***) // I take this is for creating a link for the example, so would this be if(!empty??
		}

`

**Reply by Jimmi08** [reply]
it was an example, you can put anything in that foreach, in my case there was a test if $action is printing or not.
